impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -107,6 +107,33 @@ void FenceWaiterVK::Main() {
     lock.unlock();
 
     if (terminate) {
+      // This code is copied because checking for the terminate flag after the
+      // fence waiting appears insufficient. It's possible to refactor this into
+      // a method and re-use it, but it's not clear if that might have other
+      // unintended consequences.
+      auto fences = GetFencesForWaitSet(wait_set);
+      if (fences.empty()) {
+        break;
+      }
+
+      // If there are fences to wait on, wait for them to be signaled.
+      auto device_holder = device_holder_.lock();
+      if (!device_holder) {
+        break;
+      }
+
+      const auto& device = device_holder->GetDevice();
+      auto result = device.waitForFences(
+          fences.size(),                           // fences count
+          fences.data(),                           // fences
+          false,                                   // wait for all
+          std::chrono::nanoseconds{100ms}.count()  // timeout (ns)
+      );
+      if (!(result == vk::Result::eSuccess || result == vk::Result::eTimeout)) {
+        VALIDATION_LOG << "Fence waiter encountered an unexpected error "
+                       << "while terminating.";
+      }
+
```

Comment:
This isn't quite right, because when we call wait for fences, not all of them will necessarily signal. Instead we need to also check if wait_set_ is empty - not the local copy of wait_set_.

Reply:
basically right now we have a set of N fences. We give it to waitForFences and that returns once at least one fence has updated (or timed out). But not necessarily all of the fences. So what the code is doing below is querying the fence status to determine which of the fences has updated - and remove that from the wait set. But that doesn't necessarily mean that all fences are done.

Reply:
Hmm interesting! Ok, I'll give it a look tomorrow!

Reply:
Is the change literally just `wait_set_` versus `wait_set`, or are you alluding to more?

Reply:
You are breaking too early. .waitForFences returning does not mean that all of the fences have signaled

Reply:
Reverted